sky/backends/cloud_vm_ray_backend.py

```diff
@@ -4130,6 +4132,35 @@ def post_teardown_cleanup(self,
         config = common_utils.read_yaml(handle.cluster_yaml)
         backend_utils.SSHConfigHelper.remove_cluster(handle.cluster_name)
 
+        # Confirm that instances have actually transitioned state before
+        # updating the state database.
+        attempts = 0
+        while True:
+            try:
+                logger.debug(f'instance statuses attempt {attempts + 1}')
+                node_status_dict = provision_lib.query_instances(
+                    repr(cloud),
+                    cluster_name_on_cloud,
+                    config['provider'],
+                    non_terminated_only=False)
+                # FIXME(cooperc): Some clouds (e.g. GCP) do not distinguish
+                # between "stopping/stopped" and "terminating/terminated", so we
+                # allow for either status instead of casing on `terminate`.
+                for node_id, node_status in node_status_dict.items():
+                    logger.debug(f'{node_id} status: {node_status}')
+                    if node_status not in [
+                            None, status_lib.ClusterStatus.STOPPED
+                    ]:
+                        raise RuntimeError(f'Instance {node_id} in unexpected '
+                                           f'state {node_status}.')
+                break
+            except RuntimeError:
+                attempts += 1
+                if attempts < _TEARDOWN_WAIT_MAX_ATTEMPTS:
+                    time.sleep(_TEARDOWN_WAIT_BETWEEN_ATTEMPS_SECONDS)
+                else:
+                    raise
```

Comment:
Seems this throw and catch is not necessary? Should we just use a flag?

Reply:
It won't re-throw if we should retry. Not sure what you mean by using a flag.

Reply:
Ooo, I mean exception handling is more expensive and here it seems unnecessary to raise as we can do something like the following:

```python
while True:
  is_fail = False
  for node in nodes: 
    if condition:
        is_fail = True
        break
    if is_fail:
      ...
```